Source/Libraries/CorruptCore/CorruptCoreConnector.cs

```diff
@@ -581,6 +295,305 @@ void a()
+            }
+
+            e.setReturnValue(returnList);
+        }
+
+        private static void GenerateBlastLayer(NetCoreAdvancedMessage advancedMessage, ref NetCoreEventArgs e)
+        {
+            var val = advancedMessage.objectValue as object[];
+            var sk = val[0] as StashKey;
+            var loadBeforeCorrupt = (bool)val[1];
+            var applyBlastLayer = (bool)val[2];
+            var backup = (bool)val[3];
+
+            BlastLayer bl = null;
+
+            var useSavestates = (bool)AllSpec.VanguardSpec[VSPEC.SUPPORTS_SAVESTATES];
+
+            void a()
+            {
+                lock (loadLock)
+                {
+                    //Load the game from the main thread
+                    if (useSavestates && loadBeforeCorrupt)
+                    {
+                        SyncObjectSingleton.FormExecute(() => StockpileManager_EmuSide.LoadRom_NET(sk));
+                    }
+
+                    if (useSavestates && loadBeforeCorrupt)
+                    {
+                        StockpileManager_EmuSide.LoadState_NET(sk, false);
+                    }
+
+                    bl = RtcCore.GenerateBlastLayerOnAllThreads();
+
+                    if (applyBlastLayer)
+                    {
+                        bl?.Apply(backup);
+                    }
+                }
+            }
+
+            //If the emulator uses callbacks, we do everything on the main thread and once we're done, we unpause emulation
+            if ((bool?)AllSpec.VanguardSpec[VSPEC.LOADSTATE_USES_CALLBACKS] ?? false)
+            {
+                SyncObjectSingleton.FormExecute(a);
+                e.setReturnValue(LocalNetCoreRouter.Route(NetcoreCommands.VANGUARD, NetcoreCommands.REMOTE_RESUMEEMULATION, true));
+            }
+            else //We can just do everything on the emulation thread as it'll block
+            {
+                SyncObjectSingleton.EmuThreadExecute(a, true);
+            }
+
+            if (advancedMessage.requestGuid != null)
+            {
+                e.setReturnValue(bl);
+            }
+        }
+
+        private static void ApplyBlastLayer(NetCoreAdvancedMessage advancedMessage)
+        {
+            var temp = advancedMessage.objectValue as object[];
+            var bl = (BlastLayer)temp[0];
+            var storeUncorruptBackup = (bool)temp[1];
+            var merge = (temp.Length > 2) && (bool)temp[2];
+            void a()
+            {
+                bl.Apply(storeUncorruptBackup, true, merge);
+            }
+
+            SyncObjectSingleton.EmuThreadExecute(a, true);
+        }
+
+        private static void FilterDomain(object[] objValues, ref NetCoreEventArgs e)
+        {
+            lock (loadLock)
+            {
+                var domain = (string)objValues[0];
+                var limiterListHash = (string)objValues[1];
+                var sk = objValues[2] as StashKey; //Intentionally nullable
+                var allLegalAdresses = new List<long>();
```

Comment:
This is the only helper function I had to slightly modify. I had to take `allLegalAdresses` (sic) and `e.setReturnValue(allLegalAdresses.ToArray());` out of the anonymous function because it's illegal to reference `ref NetCoreEventArgs e` inside of an anonymous function.

This is in line with how other helpers handle this. For example, `GenerateBlastLayer` has this same pattern.

Reply:
That's fine.